components/formats-gpl/src/loci/formats/in/HarmonyReader.java

```diff
@@ -0,0 +1,783 @@
+
+import ome.units.UNITS;
+import ome.units.quantity.Length;
+import ome.units.quantity.Time;
+import ome.units.unit.Unit;
+import ome.xml.model.enums.AcquisitionMode;
+import ome.xml.model.enums.EnumerationException;
+import ome.xml.model.enums.UnitsLength;
+import ome.xml.model.primitives.NonNegativeInteger;
+import ome.xml.model.primitives.PositiveFloat;
+import ome.xml.model.primitives.PositiveInteger;
+import ome.xml.model.primitives.Timestamp;
+
+import org.xml.sax.Attributes;
+
+/**
+ * HarmonyReader is the file format reader for PerkinElmer Harmony data.
+ *
+ * @author Melissa Linkert melissa at glencoesoftware.com
+ */
+public class HarmonyReader extends FormatReader {
+
+  // -- Constants --
+
+  private static final String[] XML_FILES = {"Index.idx.xml", "Index.ref.xml"};
+  private static final String MAGIC = "Harmony";
+  private static final int XML_TAG = 65500;
+
+  // -- Fields --
+
+  private Plane[][] planes;
+  private MinimalTiffReader reader;
+  private ArrayList<String> metadataFiles = new ArrayList<String>();
+
+  // -- Constructor --
+
+  /** Constructs a new Harmony reader. */
+  public HarmonyReader() {
+    super("PerkinElmer Harmony", new String[] {"xml"});
+    domains = new String[] {FormatTools.HCS_DOMAIN};
+    suffixSufficient = false;
+    hasCompanionFiles = true;
+    datasetDescription =
+      "Directory with XML file and one .tif/.tiff file per plane";
+  }
+
+  // -- IFormatReader API methods --
+
+  /* @see loci.formats.IFormatReader#getRequiredDirectories(String[]) */
+  @Override
+  public int getRequiredDirectories(String[] files)
+    throws FormatException, IOException
+  {
+    return 1;
+  }
+
+  /* @see loci.formats.IFormatReader#isSingleFile(String) */
+  @Override
+  public boolean isSingleFile(String id) throws FormatException, IOException {
+    return false;
+  }
+
+  /* @see loci.formats.IFormatReader#fileGroupOption(String) */
+  @Override
+  public int fileGroupOption(String id) throws FormatException, IOException {
+    return FormatTools.MUST_GROUP;
+  }
+
+  /* @see loci.formats.IFormatReader#isThisType(String, boolean) */
+  @Override
+  public boolean isThisType(String name, boolean open) {
+    String localName = new Location(name).getName();
+    boolean exists = false;
+    for (String XML_FILE : XML_FILES) {
+      if (localName.equals(XML_FILE)) {
+        exists = true;
+        break;
+      }
+    }
+    if (!exists) {
+      Location parent = new Location(name).getAbsoluteFile().getParentFile();
+      for (String XML_FILE : XML_FILES) {
+        Location xml = new Location(parent, XML_FILE);
+        if (xml.exists()) {
+          exists = true;
+          break;
+        }
+      }
+    }
+    if (!exists) {
+      return false;
+    }
+
+    return super.isThisType(name, open);
+  }
+
+  /* @see loci.formats.IFormatReader#isThisType(RandomAccessInputStream) */
+  @Override
+  public boolean isThisType(RandomAccessInputStream stream) throws IOException {
```

Comment:
Out of curiosity, is this right, you've implemented the option to check the xml string for "Harmony" inside a tiff image, or the Index.*.xml file? I would assume the later is the default case and would move that to the top of the method in that case.